File: src/py_examples/py_examples/slam.py
import sys
import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class SlamCommandPublisher(Node):
    def __init__(self):
        super().__init__('slam_command_publisher')
        self.publisher_ = self.create_publisher(
            String, '/integrated_command', 10)

    def publish_start_mapping(self):
        msg = String()
        msg.data = 'start_mapping'
        self.get_logger().info(f'Publishing: "{msg.data}"')
        self.publisher_.publish(msg)

    def publish_stop_mapping(self, map_name):
        msg = String()
        msg.data = f'stop_mapping:{map_name}'
        self.get_logger().info(f'Publishing: "{msg.data}"')
        self.publisher_.publish(msg)


def main(args=None):
    rclpy.init(args=args)
    slam_command_publisher = SlamCommandPublisher()

    try:
        while rclpy.ok():
            input_value = input(
                "Enter 1 to start mapping, 2 to stop mapping: ")

            if input_value == '1':
                slam_command_publisher.publish_start_mapping()
            elif input_value == '2':
                map_name = input("Enter map name: ")
                slam_command_publisher.publish_stop_mapping(map_name)
            else:
                print("Invalid input. Please enter 1 or 2.")
    except KeyboardInterrupt:
        pass
    finally:
        slam_command_publisher.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/examples/src/autonomous/slam.cpp
#include <iostream>
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>
#include <string>

// Class definition for publishing commands to control SLAM operations
class SlamCommandPublisher : public rclcpp::Node {
public:
  // Constructor for the SlamCommandPublisher class
  SlamCommandPublisher()
      : Node("slam_command_publisher") // Initialize the node with the name
                                       // "slam_command_publisher"
  {
    // Create a publisher to the "/integrated_command" topic with a queue size
    // of 10
    publisher_ = this->create_publisher<std_msgs::msg::String>(
        "/integrated_command", 10);
  }

  // Method to publish the "start_mapping" command
  void publish_start_mapping() {
    auto message = std_msgs::msg::String();
    message.data = "start_mapping";
    RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
    publisher_->publish(message);
  }

  // Method to publish the "stop_mapping" command with a specified map name
  void publish_stop_mapping(const std::string &map_name) {
    auto message = std_msgs::msg::String();
    message.data = "stop_mapping:" + map_name;
    RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
    publisher_->publish(message);
  }

private:
  // Publisher object to publish messages to the "/integrated_command" topic
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
};

int main(int argc, char *argv[]) {
  // Initialize the ROS 2 communication system
  rclcpp::init(argc, argv);
  // Create an instance of the SlamCommandPublisher node
  auto node = std::make_shared<SlamCommandPublisher>();

  int input;
  std::string map_name;

  // Infinite loop to continuously accept user input
  while (rclcpp::ok()) {
    std::cout << "Enter 1 to start mapping, 2 to stop mapping: ";
    std::cin >> input;

    if (input == 1) {
      // Publish the "start_mapping" command
      node->publish_start_mapping();
    } else if (input == 2) {
      // Prompt the user to enter the map name
      std::cout << "Enter map name: ";
      std::cin >> map_name;
      // Publish the "stop_mapping" command with the specified map name
      node->publish_stop_mapping(map_name);
    } else {
      // Handle invalid input
      std::cout << "Invalid input. Please enter 1 or 2." << std::endl;
    }
  }

  // Shutdown the ROS 2 communication system
  rclcpp::shutdown();
  return 0;
}
